File: cpp/common/bypstest/TestRemoteInlineInstance.cpp
#include "AllTests.h"


using namespace com::wilutions::byps;
using namespace com::wilutions::test;

using namespace com::wilutions::byps::test::api;
using namespace com::wilutions::byps::test::api::inl;
using namespace com::wilutions::byps::test::api::remote;

class TestRemoteInlineInstance : public TestBase {
	static BLogger log;

public:
	TestRemoteInlineInstance(void* app) : TestBase(app) {}

	PActor createActor() {
		PActor obj(new Actor());
		obj->shape = byps_ptr<BArray1<Point2D> >(new BArray1<Point2D>(3));
		for (size_t i = 0; i < obj->shape->length(); i++) {
			obj->shape->at(i) = TestUtils::createPoint2D();
		}
		obj->position._11 = 11;
		obj->position._12 = 12;
		obj->position._13 = 13;
		obj->position._21 = 21;
		obj->position._22 = 22;
		obj->position._23 = 23;
		obj->position._31 = 31;
		obj->position._32 = 32;
		obj->position._33 = 33;
		
		return obj;
	}

	void assertEquals(const std::wstring& msg, PActor value, PActor valueR) {
		assertEquals(msg + L".position", value->position, valueR->position);
		TASSERT(msg + L".shape.length", value->shape->length(), valueR->shape->length());
		for (size_t i = 0; i < value->shape->length(); i++) {
			assertEquals(msg + L".shape[i]", value->shape->at(i), valueR->shape->at(i));
		}
	}

	void assertEquals(const std::wstring& msg, const Point2D& value, const Point2D& valueR) {
		TASSERT(msg + L".x", value.x, valueR.x);
		TASSERT(msg + L".y", value.y, valueR.y);
	}

	void assertEquals(const std::wstring& msg, const Matrix2D& value, const Matrix2D& valueR) {
		TASSERT(msg + L"._xy", value._11, valueR._11);
		TASSERT(msg + L"._xy", value._12, valueR._12);
		TASSERT(msg + L"._xy", value._13, valueR._13);
		TASSERT(msg + L"._xy", value._21, valueR._21);
		TASSERT(msg + L"._xy", value._22, valueR._22);
		TASSERT(msg + L"._xy", value._23, valueR._23);
		TASSERT(msg + L"._xy", value._31, valueR._31);
		TASSERT(msg + L"._xy", value._32, valueR._32);
		TASSERT(msg + L"._xy", value._33, valueR._33);

	}

	void testRemoteInlineInstance() {

        PRemoteInlineInstance remote = client->remoteInlineInstance;

		PActor actor = createActor();
		remote->setActor(actor);
		
		assertEquals(L"actor", actor, remote->getActor());
	}

	void testRemoteInlineInstanceArray1dim() {

        PRemoteInlineInstance remote = client->remoteInlineInstance;

		byps_ptr<BArray1<Point2D> > arr(new BArray1<Point2D>(2));
		for (size_t i = 0; i < arr->length(); i++) {
			arr->at(i) = TestUtils::createPoint2D();
		}

		remote->setPoint2DArray1dim(arr);

		byps_ptr<BArray1<Point2D> > arrR = remote->getPoint2DArray1dim();
		
		TASSERT(L"arr.length", arr->length(), arrR->length());
		for (size_t i = 0; i < arr->length(); i++) {
			assertEquals(L"[i]", arr->at(i), arrR->at(i));
		}

	}

	void testRemoteInlineInstanceArray4Dim() {

        PRemoteInlineInstance remote = client->remoteInlineInstance;

		byps_ptr<BArray4<Point2D> > arr(new BArray4<Point2D>(2,3,1,1));
		for (size_t i3 = 0; i3 < arr->length3(); i3++) {
		for (size_t i2 = 0; i2 < arr->length2(); i2++) {
		for (size_t i1 = 0; i1 < arr->length1(); i1++) {
		for (size_t i0 = 0; i0 < arr->length0(); i0++) {
			arr->at(i3, i2, i1, i0) = TestUtils::createPoint2D();
		}}}}

		remote->setPoint2DArray4dim(arr);

		byps_ptr<BArray4<Point2D> > arrR = remote->getPoint2DArray4dim();
		
		TASSERT(L"arr.length3", arr->length3(), arrR->length3());
		for (size_t i3 = 0; i3 < arr->length3(); i3++) {
		TASSERT(L"arr.length2", arr->length2(), arrR->length2());
		for (size_t i2 = 0; i2 < arr->length2(); i2++) {
		TASSERT(L"arr.length1", arr->length1(), arrR->length1());
		for (size_t i1 = 0; i1 < arr->length1(); i1++) {
		TASSERT(L"arr.length0", arr->length0(), arrR->length0());
		for (size_t i0 = 0; i0 < arr->length0(); i0++) {
			assertEquals(L"arr[i]", 
				arr->at(i3, i2, i1, i0), 
				arrR->at(i3, i2, i1, i0)
			);
		}
		}
		}
		}
	}

	void testRemoteInlineInstanceList() {
		PRemoteInlineInstance remote = client->remoteInlineInstance;
		byps_ptr<std::vector<Point2D> > vec(new std::vector<Point2D>());
		vec->push_back(TestUtils::createPoint2D());
		vec->push_back(TestUtils::createPoint2D());
		remote->setPoint2DList(vec);
		byps_ptr<std::vector<Point2D> > vecR = remote->getPoint2DList();
		TASSERT(L"vec.size", vec->size(), vecR->size());
		for (unsigned i = 0; i < vec->size(); i++) {
			assertEquals(L"[i]", vec->at(i), vecR->at(i));
		}
	}

	void testRemoteInlineInstanceListList() {
		PRemoteInlineInstance remote = client->remoteInlineInstance;
		byps_ptr<std::vector<Point2D> > vec(new std::vector<Point2D>());
		vec->push_back(TestUtils::createPoint2D());
		vec->push_back(TestUtils::createPoint2D());
		byps_ptr<std::vector<byps_ptr<std::vector<Point2D> > > > vecvec(new std::vector<byps_ptr<std::vector<Point2D> > >());
		vecvec->push_back(vec);

		remote->setPoint2DListList(vecvec);
		byps_ptr<std::vector<byps_ptr<std::vector<Point2D> > > > vecvecR = remote->getPoint2DListList();
		TASSERT(L"vecvec.size", vecvec->size(), vecvecR->size());
		for (unsigned i = 0; i < vecvec->size(); i++) {
			byps_ptr<std::vector<Point2D> > vec = vecvec->at(i);
			byps_ptr<std::vector<Point2D> > vecR = vecvecR->at(i);
			TASSERT(L"vec.size", vec->size(), vecR->size());
			for (unsigned i = 0; i < vec->size(); i++) {
				assertEquals(L"[i]", vec->at(i), vecR->at(i));
			}
		}
	}

	void testRemoteInlineInstanceMap() {
		PRemoteInlineInstance remote = client->remoteInlineInstance;
		byps_ptr<std::map<int32_t, Point2D> > map(new std::map<int32_t, Point2D>());
		(*map)[11] = TestUtils::createPoint2D();
		(*map)[12] = TestUtils::createPoint2D();
		remote->setPoint2DMap(map);
		byps_ptr<std::map<int32_t, Point2D> > mapR = remote->getPoint2DMap();
		TASSERT(L"map.size", map->size(), mapR->size());
		for (std::map<int32_t, Point2D>::iterator it = map->begin(); it != map->end(); it++) {
			int32_t key = (*it).first;
			Point2D value = (*it).second;
			Point2D valueR = (*mapR)[key];
			assertEquals(L"[i]", value, valueR);
		}
	}

	virtual void init() {
		ADD_TEST(testRemoteInlineInstance);
		ADD_TEST(testRemoteInlineInstanceArray1dim);
		ADD_TEST(testRemoteInlineInstanceArray4Dim);
		ADD_TEST(testRemoteInlineInstanceList);
		ADD_TEST(testRemoteInlineInstanceListList);
		ADD_TEST(testRemoteInlineInstanceMap);
	}
};

BLogger TestRemoteInlineInstance::log("TestRemoteInlineInstance");

TestCase* TestRemoteInlineInstance_create(void* app) {
	return new TestRemoteInlineInstance(app);
}

File: java/bypstest/src/byps/test/TestRemoteInlineInstance.java
package byps.test;
/* USE THIS FILE ACCORDING TO THE COPYRIGHT RULES IN LICENSE.TXT WHICH IS PART OF THE SOURCE CODE PACKAGE */
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import byps.RemoteException;
import byps.test.api.BClient_Testser;
import byps.test.api.inl.Actor;
import byps.test.api.inl.Matrix2D;
import byps.test.api.inl.Point2D;
import byps.test.api.remote.RemoteInlineInstance;

public class TestRemoteInlineInstance {

	BClient_Testser client;
	RemoteInlineInstance remote;
	private Logger log = LoggerFactory.getLogger(TestRemoteInlineInstance.class);

	@Before
	public void setUp() throws RemoteException {
		client = TestUtilsHttp.createClient();
		remote = client.getRemoteInlineInstance();
	}
	
	@After
	public void tearDown() {
		if (client != null) {
			client.done();
		}
	}
	
	private Actor createActor() {
		Actor obj = new Actor();
		obj.position = new Matrix2D();
		
		obj.position._11 = 11;
		obj.position._12 = 12;
		obj.position._13 = 13;
		obj.position._21 = 21;
		obj.position._22 = 22;
		obj.position._23 = 23;
		obj.position._31 = 31;
		obj.position._32 = 32;
		obj.position._33 = 33;
		
		obj.shape = new Point2D[3];
		for (int i = 0; i < obj.shape.length; i++) {
			obj.shape[i] = createPoint2D();
		}
		return obj;
	}
	
	short x = 1;
	private Point2D createPoint2D() {
		Point2D p = new Point2D();
		p.x = x++;
		p.y = (short)(x*x);
		return p;
	}

	@Test
	public void testRemoteInlineInstance() throws RemoteException {
		log.info("testRemoteInlineInstance(");
		
		Actor actor = createActor();
		remote.setActor(actor);
		
		TestUtils.assertEquals(log, "actor", actor, remote.getActor());
		
		log.info(")testRemoteInlineInstance");
	}
	
	@Test
	public void testRemoteInlineInstanceArray1dim() throws RemoteException {
		log.info("testRemoteInlineInstanceArray1dim(");
		
		Point2D[] point2DArray = new Point2D[] {createPoint2D(), createPoint2D()};
		remote.setPoint2DArray1dim(point2DArray);
		
		TestUtils.assertEquals(log, "point2D", point2DArray, remote.getPoint2DArray1dim());
		
		log.info(")testRemoteInlineInstanceArray1dim");
	}

	@Test
	public void testRemoteInlineInstanceArray4dim() throws RemoteException {
		log.info("testRemoteInlineInstanceArray4dim(");
		
		Point2D[][][][] point2DArray = new Point2D[][][][] {{{{createPoint2D(), createPoint2D()}}}};
		remote.setPoint2DArray4dim(point2DArray);
		
		TestUtils.assertEquals(log, "point2D", point2DArray, remote.getPoint2DArray4dim());
		
		log.info(")testRemoteInlineInstanceArray4dim");
	}

	@Test
	public void testRemoteInlineInstanceList() throws RemoteException {
		log.info("testRemoteInlineInstanceList(");
		
		List<Point2D> point2DList = new ArrayList<Point2D>();
		point2DList.add(createPoint2D());
		point2DList.add(createPoint2D());
		remote.setPoint2DList(point2DList);
		
		TestUtils.assertEquals(log, "point2D", point2DList, remote.getPoint2DList());
		
		log.info(")testRemoteInlineInstanceList");
	}

	@Test
	public void testRemoteInlineInstanceListList() throws RemoteException {
		log.info("testRemoteInlineInstanceListList(");
		
		List<List<Point2D>> point2DListList = new ArrayList<List<Point2D>>();
		ArrayList<Point2D> point2DList = new ArrayList<Point2D>();
		point2DListList.add(point2DList);
		point2DList.add(createPoint2D());
		point2DList.add(createPoint2D());
		remote.setPoint2DListList(point2DListList);
		
		TestUtils.assertEquals(log, "point2D", point2DListList, remote.getPoint2DListList());
		
		log.info(")testRemoteInlineInstanceListList");
	}

	@Test
	public void testRemoteInlineInstanceMap() throws RemoteException {
		log.info("testRemoteInlineInstanceMap(");
		
		Map<Integer, Point2D> point2DMap = new HashMap<Integer, Point2D>();
		point2DMap.put(11, createPoint2D());
		point2DMap.put(12, createPoint2D());
		remote.setPoint2DMap(point2DMap);
		
		TestUtils.assertEquals(log, "point2D", point2DMap, remote.getPoint2DMap());
		
		log.info(")testRemoteInlineInstanceMap");
	}

}
